src/translator/html.cpp

```diff
@@ -449,13 +468,19 @@ HTML::HTML(std::string &&source, bool process_markup) {
         // separate words
         if (IsBlockElement(scanner.get_tag_name()) && !source.empty() && source.back() != ' ') source.push_back(' ');
 
-        pool_.emplace_back(new Tag{
-            scanner.get_tag_name(), std::string(),
-            IsEmtpyElement(scanner.get_tag_name())  // TODO: detect empty elements by doing a second pass and detecting
-                                                    // non-closed elements?
-        });
+        tag = new Tag{scanner.get_tag_name(), std::string(), IsEmptyElement(scanner.get_tag_name())};
```

Comment:
unique_ptr, github outage ate my nit comment, so pointing out again.

Reply:
On a closer look, turns out `pool_` is an `std::vector<std::unique_ptr>>`. This is my bad, the `Tag *` unassigned above and the `new` keyword here raised some alarms. Is it safe to assume no leak?

Would `std::make_unique<...>(...)` followed by a `std::move(...)` be more readable? Alternatively, how about we move the `new Tag{...}` block inside `emplace_back(...)` with a note `pool_` is a container holding unique-pointer?

Reply:
There is a note saying pool_ takes ownership, but I agree the previous code where `Tag *tag` didn't exist and I just referenced `bool_.back()` might have been clearer about ownership. `Tag *tag` was added to make the attribute parsing not depend on an element being on the stack.

I'll flip it around, do something like `tag = pool_.emplace_back(new Tag{}).get()` (or C++11 equivalent) again to make it clearer `pool_` owns them.

The idea of `pool_` was to have something like [kpu/preprocess/util/pool.hh:Pool](https://github.com/kpu/preprocess/blob/master/util/pool.hh#L15) eventually to save on dynamic allocations but it would be a tiny optimisation given how much string allocations I'm already doing in this bit of code anyway.

Reply:
Please excuse my lack of context regarding the details. 

If I may ask a conceptual question without knowing the full details, what is the concern here that keeps things at `std::vector<std::unique_ptr<T>>` instead of `std::vector<T>`? If it's pointer invalidation due to container growth `bool_.back()` might not be possible. If it's not pointer invalidation, you can probably get rid of the repeated allocations, I think by achieving the same with `std::vector<T>`?

Reply:
Just pointer invalidation. The `Taint` structure (which is `std::vector<Tag*>`) that's attached to each span refers to and compares tags by their address. Mainly for efficiency. And because at some point it made sense to have tags refer to other tags through parent/child relationships (but that's removed in this version). The `pool_` is there to just make sure all Tags are freed once `HTML` goes out of scope.

Technically I think I could just copy Tag structs to where they are needed, I'm currently not mutating tags after they have been assigned to a span. But that might be pretty expensive due to each tag also having a couple of strings associated with them, and as far as I understand, `std::string` is not copy-on-write.